compiler/src/iree/compiler/Preprocessing/Common/ConvertConvFilterToChannelsLast.cpp

```diff
@@ -0,0 +1,144 @@
+//===----------------------------------------------------------------------===//
+// ConvertConvFilterToChannelsLastPass
+//===----------------------------------------------------------------------===//
+
+#include "iree/compiler/Preprocessing/Common/Passes.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/Support/Debug.h"
+#include "llvm/Support/LogicalResult.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/IR/MLIRContext.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#define DEBUG_TYPE "iree-preprocessing-convert-conv-filter-to-channels-last"
+#define DBGS() (llvm::dbgs() << "[" DEBUG_TYPE "]: ")
+#define LDBG(X) LLVM_DEBUG(DBGS() << X << "\n")
+
+namespace mlir::iree_compiler::Preprocessing {
+
+#define GEN_PASS_DEF_CONVERTCONVFILTERTOCHANNELSLASTPASS
+#include "iree/compiler/Preprocessing/Common/Passes.h.inc"
+
+// Utility function to swap the last two dimensions.
+static AffineMap constructFilterMap(AffineMap map, SmallVector<int64_t> &perm) {
+  unsigned numDims = map.getNumDims();
+  ArrayRef<AffineExpr> mapResults = map.getResults();
+  SmallVector<AffineExpr, 4> exprs;
+  for (int i = 0; i < perm.size(); ++i) {
+    exprs.push_back(mapResults[perm[i]]);
+  }
+  return AffineMap::get(numDims, map.getNumSymbols(), exprs, map.getContext());
+}
+
+// Utility function to create a transpose operation.
+static Value createTransposeOp(PatternRewriter &rewriter, Location loc,
+                               Value tensor, SmallVector<int64_t> &perm) {
+  SmallVector<OpFoldResult, 4> dimSizes =
+      tensor::getMixedSizes(rewriter, loc, tensor);
+  // Dim index of H, W, F, C
+  SmallVector<OpFoldResult, 4> transposeResultDims;
+  for (int i = 0; i < perm.size(); ++i) {
+    transposeResultDims.push_back(dimSizes[perm[i]]);
+  }
+
+  auto tensorType = cast<RankedTensorType>(tensor.getType());
+  auto emptyTensor = rewriter.create<tensor::EmptyOp>(
+      loc, transposeResultDims, tensorType.getElementType());
+  return rewriter.create<linalg::TransposeOp>(loc, tensor, emptyTensor, perm)
+      .getResult()[0];
+}
+
+llvm::LogicalResult
+convertConvFilterToTargetLayout(linalg::Conv2DNhwcHwcfOp convOp,
+                                PatternRewriter &rewriter,
+                                SmallVector<int64_t> &perm) {
+  Location loc = convOp.getLoc();
+
+  // Extract operands.
+  Value input = convOp.getInputs()[0];
+  Value filter = convOp.getInputs()[1];
+  Value output = convOp.getOutputs()[0];
+
+  // Extract indexing maps.
+  AffineMap inputMap = convOp.getIndexingMapsArray()[0];
+  AffineMap filterMap = convOp.getIndexingMapsArray()[1];
+  AffineMap outputMap = convOp.getIndexingMapsArray()[2];
```

Comment:
optional nit: the comment is not useful to me because the code already says what they're doing. I think the code is already self-documenting.

https://google.github.io/styleguide/cppguide.html#Comments